### DISCUSSION TITLE
Provide a method to perform system updates

### DISCUSSION THREAD

**Discussion by Drsheppard01** [Nitrux/nitrux-bug-tracker#102] [General]
Firstly, thank you for your Linux distro, it is truly innovative. Secondly, regarding the problem itself: as I understood Nitrux doesn't use a package manager, but combines AppImage and Distrobox, and the file system is read-only, with no update mechanism to upgrade to the next version. It seems to me that overwriting the image every time is not the best user experience. I'm not a developer but conceptually, it could be 
* Updating rootfs via restic in system boot mode (plymouth)
* Downloading a lvm xfs snapshot and deploying it instead of the current one

**Comment by UriHerrera** [comment]
Hi there. Thanks, and thank you for the suggestions.

I agree that it's not the traditional Linux way of doing things. Primarily because Calamares is an installer, and I can imagine it confuses people to use an installer to upgrade. But to be fair, I do not think it's challenging to follow the steps outlined in the Notes.

The instructions in the Notes guide the user to _delete_ the root partition and create it again, not overwrite it.

Yes, it is different, but following them is not complicated. People simply aren't accustomed to using an installer without the mindset of "I'm installing X."

Regarding your suggestions, [restic is a backup program](https://restic.readthedocs.io/en/stable/010_introduction.html). The XFS filesystem has tools to manage precisely that functionality (`xfsdump` and `xfsrestore`), albeit at a partition level. 

Using LVM would present the problem that while Calamares supports LVM and the user can create an LVM group in the UI (using the Manual partition option), there's no documented way of creating an LVM group in the [modules configuration files](https://github.com/calamares/calamares/blob/calamares/src/modules/partition/partition.conf). And a quick search yields various instances where disk encryption (LUKS) doesn't work correctly with LVM in Calamares, and we just got LUKS encryption to work, so breaking that is not an option. And if the argument is that using the UI to do anything manually is not the best for user experience, we'd be at square one.

One thing I would not want either is having to restart to apply all updates, not just update the kernel; for example, Fedora and KDE Neon do this; that's Windows-like behavior, and I'm not too fond of that.

Because to me, it'd involve more complexity, i.e., booting to some "recovery state," applying updates (whatever way possible), and restarting (again).

However, what I see with using restic or LVM snapshots is that both are to create and restore backups. That being the case, we already include `bup` and the XFS tools I mentioned. We can certainly consider using either as a solution to create backups of the root during a proposed upgrade process.

Again, thank you for the suggestions. If you have more ideas on a proposed update process, please share them.

**Reply by Drsheppard01** [reply]
the point of the sentence is that since root is mounted as overlayfs and it is not modifiable, there is some "reference" root that you build. Using a backup and deduplication system will help to update and overwrite root. Conventionally, rebooting the system disables xorg-wayland, unmount root, remotely backs up from "reference" root, mounts the updated root

Programs such as [cpdd](https://github.com/selendym/cpdd) and [fsarchiver](https://github.com/fdupoux/fsarchiver) may also be useful

**Reply by UriHerrera** [reply]
The root __can be modified__; however, to make changes, the user has to _explicitly_ use `overlayroot-chroot` as we note in various topics on the Notes of these recent releases. Indeed, any changes the user makes while not using `overlayroot-chroot` only affect the upper directory in the OverlayFS structure (the default behavior), even if the user uses `sudo` or switches to the root account.

Let's see if I understand your proposal. The idea would be to provide a backup of a newer root and use one of these programs to restore that backup in an existing (at this point _older_) root.

Is that correct?

**Reply by Drsheppard01** [reply]
No, that's not quite right. The backup is not provided with the distribution, but on some server or home computer. The maintainer updates the files, then the user is synchronized as if it was when he restored the backup

**Comment by UriHerrera** [comment]
Introducing Nitrux Update System Tool or `nuts` now available here: https://github.com/Nitrux/nuts

Provides:

- Updates
- Backups
- Restores backups
